File: python/bridgetorch/bridge_torch.py
class Person:
    def __init__(self, id, timeToCrossMinutes):
        self.__id = id
        self.__timeToCrossMinutes = timeToCrossMinutes

    def __repr__(self):
        return 'Person(' + self.__id + ', ' + str(self.__timeToCrossMinutes) + ')'

    def __eq__(self, other):
        if type(self) != type(other):
            return False
        return self.__id == other.__id

    def __ne__(self, other):
        return not self.__eq__(other)

    def __cmp__(self, other):
        return cmp(self.__id, other.__id)

    def timeToCrossMinutes(self):
        return self.__timeToCrossMinutes

class TorchSide:
    LEFT = 'Left'
    RIGHT = 'Right'

    @staticmethod
    def flip(torchSide):
        if torchSide == TorchSide.LEFT:
            return TorchSide.RIGHT
        elif torchSide == TorchSide.RIGHT:
            return TorchSide.LEFT
        else:
            raise 'Unknown TorchSide ' + torchSide


class PersonsToCross:
    def __init__(self, personsToCross):
        self.__personsToCross = personsToCross

    def __repr__(self):
        return 'PersonsToCross(' + repr(self.__personsToCross) + ')'

    def contains(self, person):
        return any(person == thisPerson for thisPerson in self.__personsToCross)

    def persons(self):
        return self.__personsToCross

    def timeToCrossMinutes(self):
        timesToCrossMinutes = [person.timeToCrossMinutes() for person in self.__personsToCross]
        return max(timesToCrossMinutes)


class State:
    def __init__(self, leftSide, rightSide, torchSide, elapsedTimeMinutes = 0):
        self.__leftSide = leftSide
        self.__rightSide = rightSide
        self.__torchSide = torchSide
        self.__elapsedTimeMinutes = elapsedTimeMinutes

    def __repr__(self):
        repr = 'State(' + str(self.__leftSide)
        repr += ', ' + str(self.__rightSide)
        repr += ', ' + str(self.__torchSide)
        repr += ', ' + str(self.__elapsedTimeMinutes)
        repr += ')'
        return repr

    def nextPossibleStates(self):
        if self.__torchSide == TorchSide.LEFT:
            return self.nextPossibleStatesFromLeft()
        elif self.__torchSide == TorchSide.RIGHT:
            return self.nextPossibleStatesFromRight()
        else:
            raise 'Unknown TorchSide ' + self.__torchSide

    def nextPossibleStatesFromLeft(self):
        possiblePersonsToCross = self.possiblePersonsToCrossFromLeft()
        nextStates = [self.nextStateFromLeft(personsToCross) for personsToCross in possiblePersonsToCross]
        return nextStates

    def nextPossibleStatesFromRight(self):
        possiblePersonsToCross = self.possiblePersonsToCrossFromRight()
        nextStates = [self.nextStateFromRight(personsToCross) for personsToCross in possiblePersonsToCross]
        return nextStates

    def nextStateFromLeft(self, personsToCross):
        nextLeftSide = [person for person in self.__leftSide if not personsToCross.contains(person)]
        nextRightSide = self.__rightSide + personsToCross.persons()
        nextTorchSide = TorchSide.flip(self.__torchSide)
        nextElapsedTimeMinutes = self.__elapsedTimeMinutes + personsToCross.timeToCrossMinutes()
        return State(nextLeftSide, nextRightSide, nextTorchSide, nextElapsedTimeMinutes)

    def nextStateFromRight(self, personsToCross):
        nextLeftSide = self.__leftSide + personsToCross.persons()
        nextRightSide = [person for person in self.__rightSide if not personsToCross.contains(person)]
        nextTorchSide = TorchSide.flip(self.__torchSide)
        nextElapsedTimeMinutes = self.__elapsedTimeMinutes + personsToCross.timeToCrossMinutes()
        return State(nextLeftSide, nextRightSide, nextTorchSide, nextElapsedTimeMinutes)

    def possiblePersonsToCrossFromLeft(self):
        possiblePersonsToCross = [PersonsToCross([person1, person2])for person1 in self.__leftSide for person2 in self.__leftSide if person2 > person1]
        return possiblePersonsToCross

    def possiblePersonsToCrossFromRight(self):
        possiblePersonsToCross = [PersonsToCross([person]) for person in self.__rightSide]
        return possiblePersonsToCross

    def done(self):
        return len(self.__leftSide) == 0

    def elapsedTimeMinutes(self):
        return self.__elapsedTimeMinutes


class States:
    def __init__(self, states):
        self.__states = states

    def __repr__(self):
        return 'States(' + repr(self.__states) + ')'

    def nextStates(self):
        return States([nextState for state in self.__states for nextState in state.nextPossibleStates()])

    def finalStates(self):
        currentStates = self
        while True:
            if currentStates.allDone():
                return currentStates
            currentStates = currentStates.nextStates()

    def minTimeToCross(self):
        finalStates = self.finalStates()
        finalTimesToCross = [finalState.elapsedTimeMinutes() for finalState in finalStates.__states]
        return min(finalTimesToCross)

    def allDone(self):
        return all(state.done() for state in self.__states)


leftSide = [Person('A', 1), Person('B', 2), Person('C', 5), Person('D', 8)]
rightSide = []
torchSide = TorchSide.LEFT
initialState = State(leftSide, rightSide, torchSide)
initialStates = States([initialState])
minTimeToCross = initialStates.minTimeToCross()
print(minTimeToCross)

File: scala/bridgetorch/BridgeTorch.scala
package bridgetorch

import scala.annotation.tailrec

object BridgeTorch extends App {


  val leftSide = Set(Person("A", 1), Person("B", 2), Person("C", 5), Person("D", 8))
  val rightSide = Set[Person]()
  val torchSide = Left
  val initialState = State(leftSide, rightSide, torchSide)
  val initialStates = States(Set(initialState))
  val minTimeMinutes = initialStates.minTimeMinutes

  print(minTimeMinutes)

}

sealed trait TorchSide {
  def flip: TorchSide
}
case object Left extends TorchSide {
  override def flip = Right
}
case object Right extends TorchSide {
  override def flip = Left
}

case class Person(id: String, timeToCrossMinutes: Int) {

  override def equals(o: Any): Boolean = {
    o match {
      case other: Person => other.id.equals(id)
      case _ => false
    }
  }

  override def hashCode: Int = id.hashCode

}

case class PersonsToCross(personsToCross: Set[Person]) {

  def contains(person: Person): Boolean = {
    personsToCross.contains(person)
  }

  def timeToCrossMinutes: Int = {
    val timesToCrossMinutes = for {
      person <- personsToCross
    } yield person.timeToCrossMinutes

    timesToCrossMinutes.max
  }

}

case class State(leftSide: Set[Person], rightSide: Set[Person], torchSide: TorchSide, elapsedTimeMinutes: Int = 0) {

  def nextPossibleStates: Set[State] = {
    torchSide match {
      case Left => nextPossibleStatesFromLeft
      case Right => nextPossibleStatesFromRight
    }
  }

  def done: Boolean = leftSide.isEmpty

  private def nextPossibleStatesFromLeft: Set[State] = {
    val possiblePersonsToCross = possiblePersonsToCrossFromLeft
    val nextPossibleStates = possiblePersonsToCross.map(personsToCross => nextStateFromLeft(personsToCross))
    nextPossibleStates
  }

  private def nextPossibleStatesFromRight: Set[State] = {
    val possiblePersonsToCross = possiblePersonsToCrossFromRight
    val nextPossibleStates = possiblePersonsToCross.map(personsToCross => nextStateFromRight(personsToCross))
    nextPossibleStates
  }

  private def nextStateFromLeft(personsToCross: PersonsToCross): State = {
    val nextLeftSide = for {
    person <- leftSide
    if(!personsToCross.contains(person))
    } yield person

    val nextRightSide = rightSide ++ personsToCross.personsToCross

    val nextTorchSide = torchSide.flip

    val nextElapsedTimeMinutes = elapsedTimeMinutes + personsToCross.timeToCrossMinutes

    State(nextLeftSide, nextRightSide, nextTorchSide, nextElapsedTimeMinutes)
  }

  private def nextStateFromRight(personsToCross: PersonsToCross): State = {
    val nextLeftSide = leftSide ++ personsToCross.personsToCross
    val nextRightSide = for {
      person <- rightSide
      if(!personsToCross.contains(person))
    } yield person
    val nextTorchSide = torchSide.flip
    val nextElapsedTimeMinutes = elapsedTimeMinutes + personsToCross.timeToCrossMinutes

    State(nextLeftSide, nextRightSide, nextTorchSide, nextElapsedTimeMinutes)
  }

  private def possiblePersonsToCrossFromLeft: Set[PersonsToCross] = {
    for {
    person1 <- leftSide
    person2 <- leftSide
    if(!person1.equals(person2))
    } yield PersonsToCross(Set(person1, person2))
  }

  private def possiblePersonsToCrossFromRight: Set[PersonsToCross] = {
    for {
      person <- rightSide
    } yield PersonsToCross(Set(person))
  }

}

case class States(states: Set[State]) {

  def minTimeMinutes: Int = {
    val elapsedTimesMinutes = for {
      finalState <- finalStates.states
    } yield finalState.elapsedTimeMinutes

    elapsedTimesMinutes.min
  }

  private def nextPossibleStates: States = {
    val nextPossibleStates = states.flatMap(state => state.nextPossibleStates)
    States(nextPossibleStates)
  }

  private def finalStates: States = {
    recursiveFinalStates(this)
  }

  @tailrec
  private def recursiveFinalStates(currentStates: States): States = {
    if(currentStates.allDone) {
      return currentStates
    }

    recursiveFinalStates(currentStates.nextPossibleStates)
  }

  private def allDone: Boolean = {
    states.forall(state => state.done)
  }

}

